In [88]:
# importing pandas
import pandas as pd

#### Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [89]:
# creating DF using data from attendance fil
att = pd.read_csv('attendance.csv')

# displaying DF
att

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [90]:
# replacing letter values with numbers that correspond to each attendance value
att = att.replace('P',1)
att = att.replace('H',.5)
att = att.replace('T',.9)
att = att.replace('A',0)

# displaying DF
att

Unnamed: 0  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0      Sally           1         0.9         0.9         0.5         1.0   
1       Jane           0         1.0         0.9         0.9         0.9   
2      Billy           0         0.9         0.0         0.0         0.5   
3       John           1         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [91]:
# creating series that holds the sum of the values of each row 
total = att.sum(axis = 1)

# creating series that holds the the number of columns minus the student name column (ie. the number of days)
days = att.shape[1] - 1

# creating column that holds the value for total divided by the total number days
# ie. attendance percentage
att['percentage'] = round((total/days) * 100, 2)


att

Unnamed: 0  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0      Sally           1         0.9         0.9         0.5         1.0   
1       Jane           0         1.0         0.9         0.9         0.9   
2      Billy           0         0.9         0.0         0.0         0.5   
3       John           1         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  percentage  
0         0.0         0.9         0.9       76.25  
1         0.9         0.0         0.9       68.75  
2         0.9         1.0         0.9       52.50  
3         0.9         1.0         1.0       91.25

In [92]:
# renaming student column
att = att.rename(columns = {'Unnamed: 0': 'student'})

# creating DF that holds student name and attendance percentage
att_final = att[['student', 'percentage']]

# displaying DF
att_final

student  percentage
0   Sally       76.25
1    Jane       68.75
2   Billy       52.50
3    John       91.25

#### Read the coffee_levels.csv file.

In [94]:
# creating DF using data from coffee levels file
coffee = pd.read_csv('coffee_levels.csv')

# displaying DF
coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

#### Transform the data so that each carafe is in it's own column.

In [95]:
# using transpose to swap columns with rows so that each carafe is represented by a separate column
coffee.transpose()

0         1         2         3         4         5   \
hour                  8         9        10        11        12        13   
coffee_carafe         x         x         x         x         x         x   
coffee_amount  0.816164  0.451018  0.843279  0.335533  0.898291  0.310711   

                     6         7         8        9   ...        20       21  \
hour                 14        15        16       17  ...         8        9   
coffee_carafe         x         x         x        x  ...         z        z   
coffee_amount  0.507288  0.215043  0.183891  0.39156  ...  0.999264  0.91599   

                     22        23        24       25        26        27  \
hour                 10        11        12       13        14        15   
coffee_carafe         z         z         z        z         z         z   
coffee_amount  0.144928  0.311495  0.771947  0.39852  0.864464  0.436364   

                     28        29  
hour                 16        17  
coffee_carafe         z         z  
coffee_amount  0.280621  0.436677  

[3 rows x 30 columns]

#### Is this the best shape for the data?

In [96]:
# Answer: No because the resulting DF is too wide

#### Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [97]:
# creating DF using data from cake recipes file
cake = pd.read_csv('cake_recipes.csv')

# displaying DF
cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

#### Tidy the data as necessary.

#### Which recipe, on average, is the best? recipe b

#### Which oven temperature, on average, produces the best results? 275

#### Which combination of recipe, rack position, and temperature gives the best result? 
#### recipe b, bottom rack, 300 degrees